src/expressions/method-call-expr.md

```diff
@@ -15,22 +15,155 @@ let log_pi = pi.unwrap_or(1.0).log(2.72);
+  if the self type is `*mut T`)
+  * First for inherent methods
+  * Then for extension methods
+* And finally, a method with a `Pin` that's reborrowed, if the `pin_ergonomics`
+  feature is enabled.
+  * First for inherent methods
+  * Then for extension methods
+
+For each of those searches, if exactly one candidate is identified,
+it's picked, and the search stops. If this results in multiple possible candidates,
+then it is an error, and the receiver must be [converted][disambiguate call]
+to an appropriate receiver type to make the method call.
+
+With the example above of `SmartPtr<T>: Receiver<Target=T>`, and the receiver
+type `&SmartPtr<Foo>`, this mechanism would pick:
+
+```rust,ignore
+impl Foo {
+   fn method(self: &SmartPtr<Foo>) {}
+}
+```
 
-Then, for each candidate type `T`, search for a [visible] method with a receiver of that type in the following places:
+but would not pick
 
-1. `T`'s inherent methods (methods implemented directly on `T`).
-1. Any of the methods provided by a [visible] trait implemented by `T`.
-   If `T` is a type parameter, methods provided by trait bounds on `T` are looked up first.
-   Then all remaining methods in scope are looked up.
+```rust,ignore
+impl Foo {
+   fn method(self: &Foo) {}
+}
+```
 
-> Note: the lookup is done for each type in order, which can occasionally lead to surprising results.
+because the receiver could not be converted to `&Foo` using the `Deref` chain,
+only the `Receiver` chain.
+
+## Extra details
+
+There are a few details not considered in this overview:
+
+* The search for candidate methods will also consider searches for
+  incoherent types if `rustc_has_incoherent_inherent_impls` is active for
+  a `dyn`, struct, enum, or foreign type.
+* If there are multiple candidates from traits, they may in fact be
+  identical, and the picking operation collapses them to a single pick to avoid
+  reporting conflicts.
+* Extra searches are performed to spot "shadowing" of pointee methods
+  by smart pointer methods, during the picking process. If a by-value pick
+  is going to be returned, an extra search is performed for a `&T` or
+  `&mut T` method. Similarly, if a `&T` method is to be returned, an extra
+  search is performed for `&mut T` methods. These extra searches consider
+  only inherent methods, where `T` is identical, but the method is
+  found from a step further along the `Receiver` chain. If any such method
+  is found, an ambiguity error is emitted.
+* An error is emitted if we reached a recursion limit.
+* The picking process emits some adjustments which must be made to the
+  receiver type in order to get to the correct `self` type. This includes
+  a number of dereferences, a possible autoreferencing, a conversion from
+  a mutable pointer to a constant pointer, or a pin reborrow.
+* Extra lists are maintained for diagnostic purposes:
+  unstable candidates, unsatisfied predicates, and static candidates.
```

Comment:
Suggestion: Maybe leave notes about diagnostics out of the overview if saying under "extra details" that they're left out?

Reply:
Done, at least if I understand you correctly - I've moved the remaining diagnostic-related point down to the Extra Details list.